common/common.go

```diff
@@ -0,0 +1,218 @@
+	result, rpcErr := rpcClient.RawRequest("getblockchaininfo", make([]json.RawMessage, 0))
+
+	var err error
+	var errCode int64
+
+	// For some reason, the error responses are not JSON
+	if rpcErr != nil {
+		errParts := strings.SplitN(rpcErr.Error(), ":", 2)
+		errCode, err = strconv.ParseInt(errParts[0], 10, 32)
+		// Check to see if we are requesting a height the zcashd doesn't have yet
+		if err == nil && errCode == -8 {
+			return -1, -1, "", "", nil
+		}
+		return -1, -1, "", "", errors.Wrap(rpcErr, "error requesting block")
+	}
+
+	var f interface{}
+	err = json.Unmarshal(result, &f)
+	if err != nil {
+		return -1, -1, "", "", errors.Wrap(err, "error reading JSON response")
+	}
+
+	chainName := f.(map[string]interface{})["chain"].(string)
+
+	upgradeJSON := f.(map[string]interface{})["upgrades"]
+	saplingJSON := upgradeJSON.(map[string]interface{})["76b809bb"] // Sapling ID
+	saplingHeight := saplingJSON.(map[string]interface{})["activationheight"].(float64)
+
+	blockHeight := f.(map[string]interface{})["headers"].(float64)
+
+	consensus := f.(map[string]interface{})["consensus"]
+	branchID := consensus.(map[string]interface{})["nextblock"].(string)
+
+	return int(saplingHeight), int(blockHeight), chainName, branchID, nil
+}
+
+func getBlockFromRPC(rpcClient *rpcclient.Client, height int) (*walletrpc.CompactBlock, error) {
+	params := make([]json.RawMessage, 2)
+	params[0] = json.RawMessage("\"" + strconv.Itoa(height) + "\"")
+	params[1] = json.RawMessage("0")
+	result, rpcErr := rpcClient.RawRequest("getblock", params)
+
+	var err error
+	var errCode int64
+
+	// For some reason, the error responses are not JSON
+	if rpcErr != nil {
+		errParts := strings.SplitN(rpcErr.Error(), ":", 2)
+		errCode, err = strconv.ParseInt(errParts[0], 10, 32)
+		// Check to see if we are requesting a height the zcashd doesn't have yet
+		if err == nil && errCode == -8 {
+			return nil, nil
+		}
+		return nil, errors.Wrap(rpcErr, "error requesting block")
+	}
+
+	var blockDataHex string
+	err = json.Unmarshal(result, &blockDataHex)
+	if err != nil {
+		return nil, errors.Wrap(err, "error reading JSON response")
+	}
+
+	blockData, err := hex.DecodeString(blockDataHex)
+	if err != nil {
+		return nil, errors.Wrap(err, "error decoding getblock output")
+	}
+
+	block := parser.NewBlock()
+	rest, err := block.ParseFromSlice(blockData)
+	if err != nil {
+		return nil, errors.Wrap(err, "error parsing block")
+	}
+	if len(rest) != 0 {
+		return nil, errors.New("received overlong message")
+	}
+
+	return block.ToCompact(), nil
+}
+
+func BlockIngestor(rpcClient *rpcclient.Client, cache *BlockCache, log *logrus.Entry,
+	stopChan chan bool, startHeight int) {
+	reorgCount := 0
+	height := startHeight
+	timeoutCount := 0
+
+	// Start listening for new blocks
+	for {
+		select {
+		case <-stopChan:
+			break
+
+		case <-time.After(15 * time.Second):
+			for {
+				if reorgCount > 0 {
+					height -= 10
```

Comment:
This can set `height` to a value less than `startHeight`. If this happens, then the cache can contain invalid blocks, because the cache's `Add` function only clears higher-height cached blocks if the block being added falls within the range of cached blocks (if `height < startHeight`, then it won't fall in range, so the reorged-away blocks don't get cleared). We can do: `height = Max(height - 10, startHeight);` here to prevent this.